### PR TITLE
Add tests for learner read

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #7256

Problem Summary:
1. Add tests for learner read and RegionRangeIndex
1. Remove RegionMockTest from non-test utils
1. Remove obsolete codes, and integrate similar codes

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Transaction/RegionTable.h [hunk 1]
```diff
@@ -231,7 +231,7 @@ class RegionTable : private boost::noncopyable
 
     FlushThresholds flush_thresholds;
 
-    Context * const context;
+    Context * context;
```
Comment (CalvinNeo) [inline dbms/src/Storages/Transaction/RegionTable.h:234]: I will revert this later

**Review by hongyunyan — APPROVED** [review]

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM

File: dbms/src/Storages/Transaction/LearnerRead.cpp [hunk 2]
```diff
@@ -457,6 +459,7 @@ void validateQueryInfo(
         if (auto iter = regions_snapshot.find(region_query_info.region_id); //
             iter == regions_snapshot.end() || iter->second != region)
         {
+            // If snapshot is applied during learner read, we should abort.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/LearnerRead.cpp:462]: what do you mean "abort" here? abort the tiflash process or throw an exception to abort the read request?

Comment (CalvinNeo) [reply dbms/src/Storages/Transaction/LearnerRead.cpp:462]: I will make this clear

**Pull request merged**